packages/event-handler/src/rest/RouteHandlerRegistry.ts

```diff
@@ -0,0 +1,53 @@
+import type { GenericLogger } from '@aws-lambda-powertools/commons/types';
+import type { RouteRegistryOptions } from '../types/rest.js';
+import type { Route } from './Route.js';
+import { validatePathPattern } from './utils.js';
+
+class RouteHandlerRegistry {
+  readonly #routes: Map<string, Route> = new Map();
+  readonly #routesByMethod: Map<string, Route[]> = new Map();
+
+  readonly #logger: Pick<GenericLogger, 'debug' | 'warn' | 'error'>;
+
+  constructor(options: RouteRegistryOptions) {
+    this.#logger = options.logger;
+  }
+
+  public register(route: Route): void {
+    const { isValid, issues } = validatePathPattern(route.path);
+    if (!isValid) {
+      for (const issue of issues) {
+        this.#logger.warn(issue);
+      }
+      return;
+    }
+
+    if (this.#routes.has(route.id)) {
+      this.#logger.warn(
+        `Handler for method: ${route.method} and path: ${route.path} already exists. The previous handler will be replaced.`
+      );
+    }
+
+    this.#routes.set(route.id, route);
+
+    if (!this.#routesByMethod.has(route.method)) {
+      this.#routesByMethod.set(route.method, []);
+    }
+    // biome-ignore lint/style/noNonNullAssertion: Map.set operation above ensures Map.get won't return undefined
+    this.#routesByMethod.get(route.method)!.push(route);
+  }
```

Comment:
If we change the flow to something like this, we can avoid the non-null assertion:

```ts
const routesByMethod =
this.#routesByMethod.get(route.method) || [];
routesByMethod.push(route);
this.#routesByMethod.set(route.method, routesByMethod);
```

Additionally, if we make `routesByMethod` a set `new Set<Route>()` we can avoid duplicates. If we did that, the `getRoutesByMethod` method would become and tests would still pass:

```ts
public getRoutesByMethod(method: string): Route[] {
  const routes = this.#routesByMethod.get(method.toUpperCase());
  if (!routes) {
    return [];
  }
  return Array.from(routes);
}
```

Reply:
Ah nice sugggestion for the non-null. We'll have an even nicer solution once `Map.getOrInsert` arrives: https://github.com/tc39/proposal-upsert.

I like the set idea too.